main/src/library/Fixpoint/Solver.flix

```diff
@@ -87,42 +88,57 @@ namespace Fixpoint {
-        let getValue = term -> match term {
-            case HeadTerm.Lit(v, _) => v
-            case _                  => bug!("Unexpected non-value term.")
-        };
-        let unbox = terms -> (Boxable.unbox(getValue(terms[0])), Boxable.unbox(getValue(terms[1])));
-        let Datalog(facts, _) = d;
-        let headPredicates = restrict(facts, Set#{p}) |> Array.map(match Constraint(head, _, _) -> head);
-        let headAtoms = headPredicates |> Array.map(match HeadAtom(_, _, terms, _) -> unbox(terms));
-        headAtoms
-    } as & Pure
+    pub def facts2(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2)] with Boxable[t1], Boxable[t2] =
+        let f = terms -> (unbox(terms[0]), unbox(terms[1]));
+        (factsOf(p, d) |> Array.map(f)) as & Pure
 
     ///
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
-    pub def facts3(_p: PredSym, _d: Datalog[v]): Array[(t1, t2, t3)] with Boxable[t1], Boxable[t2], Boxable[t3] =
-        [] as & Pure // TODO: Benj: Implement
+    pub def facts3(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2, t3)] with Boxable[t1], Boxable[t2], Boxable[t3] =
+        let f = terms -> (unbox(terms[0]), unbox(terms[1]), unbox(terms[2]));
+        (factsOf(p, d) |> Array.map(f)) as & Pure
 
     ///
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
-    pub def facts4(_p: PredSym, _d: Datalog[v]): Array[(t1, t2, t3, t4)] with Boxable[t1], Boxable[t2], Boxable[t3], Boxable[t4] =
-        [] as & Pure // TODO: Benj: Implement
+    pub def facts4(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2, t3, t4)] with Boxable[t1], Boxable[t2], Boxable[t3], Boxable[t4] =
+        let f = terms -> (unbox(terms[0]), unbox(terms[1]), unbox(terms[2]), unbox(terms[3]));
+        (factsOf(p, d) |> Array.map(f)) as & Pure
 
     ///
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
-    pub def facts5(_p: PredSym, _d: Datalog[v]): Array[(t1, t2, t3, t4, t5)] with Boxable[t1], Boxable[t2], Boxable[t3], Boxable[t4], Boxable[t5] =
-        [] as & Pure // TODO: Benj: Implement
+    pub def facts5(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2, t3, t4, t5)] with Boxable[t1], Boxable[t2], Boxable[t3], Boxable[t4], Boxable[t5] =
+        let f = terms -> (unbox(terms[0]), unbox(terms[1]), unbox(terms[2]), unbox(terms[3]), unbox(terms[4]));
+        (factsOf(p, d) |> Array.map(f)) as & Pure
+
+    ///
+    /// Returns an array of facts associated with the given predicate symbol `p` in the given Datalog program `d`.
+    ///
+    def factsOf(p: PredSym, d: Datalog[v]): Array[Array[v]] & Impure =
+        let Datalog(facts, _) = d;
+        restrict(facts, Set#{p}) |>
+        Array.map(match Constraint(head, _, _) -> head) |>
+        Array.map(match HeadAtom(_, _, terms, _) -> terms) |>
+        Array.map(terms -> terms |> Array.map(headTermValue))
```

Comment:
I think this could be replaced by a call to `factToTuple`.

Reply:
Feel free to simplify that